### PR TITLE
Add indenting formatter rules for bindings and nested module declarations

### PR DESCRIPTION
**Pull request by reacheight**
Added some formatter rules for indents inside
- bindings
- nested modules

Also added registry value to enable formatter.

### CODE REVIEW

File: rider-fsharp/src/main/resources/META-INF/plugin.xml [hunk 1]
```diff
@@ -84,6 +84,7 @@
     </intentionAction>
 
     <registryKey key="rider.fsharp.experimental" description="Enable experimental F# features" defaultValue="false" restartRequired="false"/>
+    <registryKey key="rider.fsharp.formatter" description="Enable new F# code formatter" defaultValue="false" restartRequired="false"/>
```
Comment (auduchinok) [inline rider-fsharp/src/main/resources/META-INF/plugin.xml]: We didn't have a working R# formatter previously, and Fantomas is doing a somewhat different thing in Reformat Code action, so I'd prefer it without `new` word here. Though, I doubt many people will ever see this temporary setting. :)

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatterInfoProvider.cs [hunk 2]
```diff
@@ -12,8 +15,78 @@ public class FSharpFormatterInfoProvider :
...
+        ("ForEachExpr", ElementType.FOR_EACH_EXPR, ForEachExpr.DO_EXPR),
+        ("WhileExpr", ElementType.WHILE_EXPR, WhileExpr.DO_EXPR),
+        ("DoExpr", ElementType.DO_EXPR, DoExpr.EXPR),
+        ("AssertExpr", ElementType.ASSERT_EXPR, AssertExpr.EXPR),
+        ("LazyExpr", ElementType.LAZY_EXPR, LazyExpr.EXPR),
+        ("FixedExpr", ElementType.FIXED_EXPR, FixedExpr.EXPR),
+        ("InferredUpcastExpr", ElementType.INFERRED_UPCAST_EXPR, InferredUpcastExpr.EXPR),
+        ("InferredDowncastExpr", ElementType.INFERRED_DOWNCAST_EXPR, InferredDowncastExpr.EXPR),
+        ("ComputationExpr", ElementType.COMPUTATION_EXPR, ComputationExpr.EXPR),
+        ("LambdaExpr", ElementType.LAMBDA_EXPR, LambdaExpr.EXPR),
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatterInfoProvider.cs]: Does it support deindenting like in the following code?

```fsharp
let foo x = x |> fun _ ->
    ()
```

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatterInfoProvider.cs [hunk 3]
```diff
@@ -12,8 +15,78 @@ public class FSharpFormatterInfoProvider :
...
+      var bindingAndModuleDeclIndentingRulesParameters = new[]
+      {
+        ("NestedModuleDeclaration", ElementType.NESTED_MODULE_DECLARATION, NestedModuleDeclaration.MODULE_MEMBER),
+        ("TopBinding", ElementType.TOP_BINDING, TopBinding.CHAMELEON_EXPR),
+        ("LocalBinding", ElementType.LOCAL_BINDING, LocalBinding.EXPR),
+      };
+
+      var synExprIndentingRulesParameters = new[]
+      {
+        ("ForExpr", ElementType.FOR_EXPR, ForExpr.DO_EXPR),
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatterInfoProvider.cs]: Would be great to have some simple tests like checking if the following code is unchanged.

```fsharp
for _ in () do
    ()
```

File: ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatterInfoProvider.cs (hunk 2 above)
Comment (reacheight) [reply ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpCodeFormatterInfoProvider.cs]: Yes.

File: ReSharper.FSharp/src/FSharp.Common/src/Util/RegistryUtil.fs [hunk 4]
```diff
@@ -8,19 +8,22 @@ open JetBrains.ReSharper.Psi.Tree
 open JetBrains.Rider.Model
 
 module FSharpRegistryUtil =
-    type AllowExperimentalFeaturesCookie() =
+    [<AbstractClass>]
+    type EnabledCookieBase<'T when 'T :> EnabledCookieBase<'T> and 'T : (new : unit -> 'T)>() =
         static let mutable enabled = false
```
[the remark this answers is not in the document]
Comment (reacheight) [reply ReSharper.FSharp/src/FSharp.Common/src/Util/RegistryUtil.fs]: It's unique for every combination of type arguments, when the base class is generic.
изображение

[the remark this answers is not in the document]
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp.Common/src/Util/RegistryUtil.fs]: OK, I see, thanks!

**Pull request merged**